File: apps/web/lib/api/conditional-order.ts
/**
 * 条件单 API client(ADR 0041 刀3)· 照 virtual.ts authed 范式。
 *
 * 三端点(刀1 已上生产):
 *   POST   /api/v1/virtual/conditional-orders        挂单(LIMIT / SL / TP)
 *   GET    /api/v1/virtual/conditional-orders?status= 列表(仅本人 · 倒序)
 *   DELETE /api/v1/virtual/conditional-orders/{id}    撤单(仅 ACTIVE → 204)
 *
 * 金额字段全 string(房规:后端 Decimal 序列化 string,避免 JS 浮点损失)。
 * 🔴 红线:挂/撤/看,成交由后端 60s 扫描器走唯一虚拟撮合入口 · 前端绝不撮合。
 */

import type { ConditionalKind, ConditionalStatus } from '@/lib/conditional'
import type { OrderSide, PositionSide } from '@/lib/api/virtual'
import type { Market } from '@midas/shared'

const API_BASE = process.env.NEXT_PUBLIC_API_URL ?? 'http://localhost:8000'

export interface ConditionalOrderCreate {
  symbol: string
  market: Market
  order_kind: ConditionalKind
  side: OrderSide
  /** 不传 = long(后端默认)· 仅空头仓的 SL/TP 传 short */
  position_side?: PositionSide
  trigger_price: string
  /** LIMIT 必填;SL/TP 省略 = 触发时平全仓 */
  quantity?: string
}

export interface ConditionalOrder {
  id: number
  symbol: string
  market: string
  order_kind: ConditionalKind
  side: OrderSide
  position_side: PositionSide
  trigger_price: string
  quantity: string | null
  status: ConditionalStatus
  triggered_order_id: number | null
  note: string | null
  created_at: string
  updated_at: string
}

export class ConditionalApiError extends Error {
  constructor(
    public readonly status: number,
    public readonly detail: string,
  ) {
    super(`ConditionalApi ${status}: ${detail}`)
    this.name = 'ConditionalApiError'
  }
}

async function readDetail(resp: Response): Promise<string> {
  try {
    const body = (await resp.json()) as { detail?: unknown }
    return typeof body.detail === 'string' ? body.detail : `HTTP ${resp.status}`
  } catch {
    return `HTTP ${resp.status}`
  }
}

function authHeaders(token: string): HeadersInit {
  return {
    Authorization: `Bearer ${token}`,
    'Content-Type': 'application/json',
  }
}

export async function postConditionalOrder(
  token: string, input: ConditionalOrderCreate,
): Promise<ConditionalOrder> {
  const r = await fetch(`${API_BASE}/api/v1/virtual/conditional-orders`, {
    method: 'POST',
    headers: authHeaders(token),
    body: JSON.stringify(input),
  })
  if (!r.ok) throw new ConditionalApiError(r.status, await readDetail(r))
  return (await r.json()) as ConditionalOrder
}

export async function listConditionalOrders(
  token: string, opts: { status?: ConditionalStatus } = {}, signal?: AbortSignal,
): Promise<ConditionalOrder[]> {
  const qs = opts.status ? `?status=${opts.status}` : ''
  const r = await fetch(`${API_BASE}/api/v1/virtual/conditional-orders${qs}`, {
    headers: authHeaders(token), signal,
  })
  if (!r.ok) throw new ConditionalApiError(r.status, await readDetail(r))
  return (await r.json()) as ConditionalOrder[]
}

export async function cancelConditionalOrder(token: string, id: number): Promise<void> {
  const r = await fetch(`${API_BASE}/api/v1/virtual/conditional-orders/${id}`, {
    method: 'DELETE',
    headers: authHeaders(token),
  })
  if (!r.ok) throw new ConditionalApiError(r.status, await readDetail(r))
}

File: apps/api/app/schemas/conditional_order.py
"""条件单 Pydantic 契约(ADR 0041 刀1)。

🔴 红线:挂单簿管理 API(挂/撤/列)—— 本层不撮合不成交;触发成交是刀2 扫描器,
   届时仍走 place_market_order 唯一入口。本刀条件单【不会触发】(有意分界)。
"""

from __future__ import annotations

from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field

from app.models.conditional_order import ConditionalKind, ConditionalStatus
from app.models.virtual import OrderSide, PositionSide
from app.schemas.market import Market


class ConditionalOrderCreate(BaseModel):
    """POST /virtual/conditional-orders 入参。

    - LIMIT:quantity 必填(端点校验);side=BUY 低吸 / SELL 高抛。
    - STOP_LOSS / TAKE_PROFIT:持仓后独立挂(ADR 未决②定型为 (a))· quantity 可空=平全仓 ·
      side 必须是该仓向的平仓方向(LONG→SELL / SHORT→BUY,端点校验防脏数据)。
    - crypto 语义=perp(symbol 用 Binance 风格 BTCUSDT);spot 各市场用现有符号形态。
    """

    model_config = ConfigDict(extra="forbid")

    symbol: str = Field(min_length=1, max_length=64)
    market: Market
    order_kind: ConditionalKind
    side: OrderSide
    position_side: PositionSide = PositionSide.LONG
    trigger_price: Decimal = Field(gt=0)
    quantity: Decimal | None = Field(default=None, gt=0)


class ConditionalOrderResponse(BaseModel):
    """单张条件单(列表项与详情同型 · 字段轻,无需精简版)。"""

    model_config = ConfigDict(extra="forbid", frozen=True, from_attributes=True)

    id: int
    symbol: str
    market: str
    order_kind: ConditionalKind
    side: OrderSide
    position_side: PositionSide
    trigger_price: Decimal
    quantity: Decimal | None
    status: ConditionalStatus
    triggered_order_id: int | None
    note: str | None
    created_at: datetime
    updated_at: datetime
